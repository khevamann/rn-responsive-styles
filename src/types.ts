export enum DEVICE_SIZES {
  EXTRA_SMALL_DEVICE = 'extra_small',
  SMALL_DEVICE = 'small',
  MEDIUM_DEVICE = 'medium',
  LARGE_DEVICE = 'large',
  EXTRA_LARGE_DEVICE = 'extra_large',
}
