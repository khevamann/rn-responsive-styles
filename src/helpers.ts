import { DEVICE_SIZES } from './types';

export function deviceSize(width: number) {
  if (width > 1200) return DEVICE_SIZES.EXTRA_LARGE_DEVICE;
  if (width > 992) return DEVICE_SIZES.LARGE_DEVICE;
  if (width > 768) return DEVICE_SIZES.MEDIUM_DEVICE;
  if (width > 540) return DEVICE_SIZES.SMALL_DEVICE;
  return DEVICE_SIZES.EXTRA_SMALL_DEVICE;
}
