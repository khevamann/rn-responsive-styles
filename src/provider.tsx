import * as React from 'react'

export type Breakpoints = [number, number, number, number]

const BreakpointsContext = React.createContext<Breakpoints>([1200, 992, 768, 540])

export const useBreakpoints = () => React.useContext(BreakpointsContext)

// Provider to allow for passing in custom props to control Responsive Styles
type Props = {
  children: React.ReactNode
  breakpoints: Breakpoints
}

export function BreakpointsProvider({ children, breakpoints }: Props) {
  return <BreakpointsContext.Provider value={breakpoints}>{children}</BreakpointsContext.Provider>
}
