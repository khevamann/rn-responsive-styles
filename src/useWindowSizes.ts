import { useEffect, useState } from 'react'
import { Dimensions, ScaledSize } from 'react-native'
import { DEVICE_SIZES } from './types'

/**
 * Hook to watch changes in screenSize and report the dimensions whenever the device size changes
 * @param breakpoints a function that returns the device size given a width
 */
export default function useWindowSizes(breakpoints: (width: number) => DEVICE_SIZES) {
  const [dims, setDims] = useState(() => Dimensions.get('window'))

  useEffect(() => {
    function handleChange({ window }: { window: ScaledSize }) {
      // Only update the dimensions when the device size changes
      setDims((prev) => (breakpoints(prev.width) === breakpoints(window.width) ? prev : window))
    }

    const listener = Dimensions.addEventListener('change', handleChange)
    // We might have missed an update between calling `get` in render and
    // `addEventListener` in this handler, so we set it here. If there was
    // no change, React will filter out this update as a no-op.
    setDims(Dimensions.get('window'))
    return listener && listener.remove
  }, [])

  return dims
}
