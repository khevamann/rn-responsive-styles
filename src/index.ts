import useDeviceSize from './hooks/useDeviceSize'
import { DEVICE_SIZES } from './types'
import { maxSize, minSize } from './helpers'
import { BreakpointsProvider } from './provider'
import CreateResponsiveStyle from './createResponsiveStyle'

export { CreateResponsiveStyle, BreakpointsProvider, useDeviceSize, DEVICE_SIZES, minSize, maxSize }
