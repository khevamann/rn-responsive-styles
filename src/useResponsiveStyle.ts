import { StyleSheet } from 'react-native'
import { deviceSize } from './helpers'
import useWindowSizes from './useWindowSizes'

export default function useResponsiveStyle<T>(styles: StyleSheet.NamedStyles<any>, provided_sizes) {
  return () => {
    const layout = useWindowSizes(deviceSize, provided_sizes)
    const size = deviceSize(layout.width, provided_sizes);    
    
    return {
      styles: (style: keyof T) => StyleSheet.compose(styles[style], styles[`${size}_${style}`]),
      deviceSize: size,
    }
  }
}
