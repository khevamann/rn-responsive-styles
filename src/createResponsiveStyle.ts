import { DEVICE_SIZES, MEDIA_QUERY } from './types'
import { ImageStyle, StyleSheet, TextStyle, ViewStyle } from 'react-native'
import useResponsiveStyle from './hooks/useResponsiveStyle'

type ReturnType<GlobalStyles> = () => GlobalStyles

function CreateResponsiveStyle<DefaultStyles extends StyleSheet.NamedStyles<any>, OverrideStyles extends DefaultStyles>(
  webStyles: DefaultStyles,
  styleOverrides?: Record<DEVICE_SIZES | MEDIA_QUERY, Partial<StyleSheet.NamedStyles<OverrideStyles>>>
): ReturnType<DefaultStyles> {
  const overrides: StyleSheet.NamedStyles<any> = {}

  // Create custom style names based on the device overrides if provided
  if (styleOverrides) {
    Object.entries(styleOverrides).forEach(([key, value]) => {
      Object.entries(value).forEach(([className, value2]) => {
        overrides[`$$${key}$$_${className}`] = value2 as ViewStyle | TextStyle | ImageStyle
      })
    })
  }

  // Merge the stylesheets example:
  // {
  //   container: {
  //     color: 'red'
  //   },
  //   $$large$$_container: {
  //     color: 'blue'
  //   },
  //   $$extrasmall+small$$_container: {
  //     color: 'green'
  //   },
  // }
  const styles = StyleSheet.create<DefaultStyles>({ ...webStyles, ...overrides })

  return useResponsiveStyle<DefaultStyles>(styles)
}

export default CreateResponsiveStyle
