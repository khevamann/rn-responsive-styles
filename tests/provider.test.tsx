import * as React from 'react'
import { ScaledSize } from 'react-native'
import { cleanup, render } from '@testing-library/react-native'
import Sample from './Sample'
import { BreakpointsProvider } from '../src'

afterEach(cleanup)

const mockDimensions = ({ width }: Pick<ScaledSize, 'width'>) => {
  jest.resetModules()
  jest.doMock('react-native/Libraries/Utilities/Dimensions', () => ({
    get: jest.fn().mockReturnValue({ width }),
    addEventListener: jest.fn(),
  }))
}
describe('tests that device sizes render appropriately', () => {
  test('tests that with breakpoints this will be a large device', () => {
    mockDimensions({ width: 414 })
    const tree = render(
      <BreakpointsProvider breakpoints={[500, 400, 300, 200]}>
        <Sample />
      </BreakpointsProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot('large')
  })
  test('tests that with breakpoints this will be a extra small device', () => {
    mockDimensions({ width: 150 })
    const tree = render(
      <BreakpointsProvider breakpoints={[500, 400, 300, 200]}>
        <Sample />
      </BreakpointsProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot('extrasmall')
  })
  test('tests that with breakpoints this will be a medium device', () => {
    mockDimensions({ width: 399 })
    const tree = render(
      <BreakpointsProvider breakpoints={[500, 400, 300, 200]}>
        <Sample />
      </BreakpointsProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot('medium')
  })
})
