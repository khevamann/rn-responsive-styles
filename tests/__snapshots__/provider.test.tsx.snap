// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests that SSR works correctly tests that SSR will not block on iOS 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "purple",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Device Size: 
    extrasmall
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Color: 
    purple
  </Text>
  <Text>
    Only rendered for devices smaller than medium
  </Text>
</View>
`;

exports[`tests that SSR works correctly tests that SSR will not block on web 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "purple",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Device Size: 
    extrasmall
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Color: 
    purple
  </Text>
  <Text>
    Only rendered for devices smaller than medium
  </Text>
</View>
`;

exports[`tests that SSR works correctly tests that SSR will show loading while rendering on the server 1`] = `
<Text>
  Loading...
</Text>
`;

exports[`tests that device sizes render appropriately tests that with breakpoints this will be a extra small device: extrasmall 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "purple",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Device Size: 
    extrasmall
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Color: 
    purple
  </Text>
  <Text>
    Only rendered for devices smaller than medium
  </Text>
</View>
`;

exports[`tests that device sizes render appropriately tests that with breakpoints this will be a large device: large 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "purple",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "blue",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 30,
        },
      ]
    }
  >
    Device Size: 
    large
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 30,
        },
      ]
    }
  >
    Color: 
    blue
  </Text>
  <Text>
    Only rendered for devices larger than medium
  </Text>
</View>
`;

exports[`tests that device sizes render appropriately tests that with breakpoints this will be a medium device: medium 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "purple",
        "flex": 1,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "green",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Device Size: 
    medium
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "white",
        },
        Object {
          "fontSize": 20,
        },
      ]
    }
  >
    Color: 
    green
  </Text>
  <Text>
    Only rendered for medium devices
  </Text>
</View>
`;
